oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/migration/version/VersionCopierTest.java

```diff
@@ -0,0 +1,167 @@
+import org.apache.jackrabbit.oak.api.Tree;
+import org.apache.jackrabbit.oak.api.Type;
+import org.apache.jackrabbit.oak.commons.PathUtils;
+import org.apache.jackrabbit.oak.core.ImmutableRoot;
+import org.apache.jackrabbit.oak.namepath.NamePathMapper;
+import org.apache.jackrabbit.oak.plugins.memory.MemoryNodeStore;
+import org.apache.jackrabbit.oak.plugins.migration.NodeStateCopier;
+import org.apache.jackrabbit.oak.plugins.version.ReadOnlyVersionManager;
+import org.apache.jackrabbit.oak.plugins.version.ReadWriteVersionManager;
+import org.apache.jackrabbit.oak.spi.state.NodeBuilder;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.apache.jackrabbit.oak.spi.state.NodeStore;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import static java.util.Collections.singletonList;
+import static org.apache.jackrabbit.JcrConstants.JCR_FROZENNODE;
+import static org.apache.jackrabbit.JcrConstants.JCR_MIXINTYPES;
+import static org.apache.jackrabbit.JcrConstants.JCR_PRIMARYTYPE;
+import static org.apache.jackrabbit.JcrConstants.JCR_UUID;
+import static org.apache.jackrabbit.JcrConstants.MIX_VERSIONABLE;
+import static org.apache.jackrabbit.JcrConstants.NT_UNSTRUCTURED;
+import static org.apache.jackrabbit.oak.InitialContentHelper.INITIAL_CONTENT;
+import static org.apache.jackrabbit.oak.InitialContentHelper.INITIAL_CONTENT_FROZEN_NODE_REFERENCEABLE;
+import static org.apache.jackrabbit.oak.plugins.migration.NodeStateTestUtils.commit;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+public class VersionCopierTest {
+
+    @Test
+    public void copyVersionSourceFrozenNodeReferenceable() throws Exception {
+        String path = "/foo";
+        NodeStore source = createVersionFor(path, createStore(true));
+        NodeStore target = createStore(false);
+
+        copyContent(source, target, path);
+        copyVersionStorage(source, target);
+        assertVersionNotReferenceable(target.getRoot(), path);
+    }
+
+    @Test
+    public void copyVersionSourceFrozenNodeNotReferenceable() throws Exception {
+        String path = "/foo";
+        NodeStore source = createVersionFor(path, createStore(false));
+        NodeStore target = createStore(false);
+
+        copyContent(source, target, path);
+        copyVersionStorage(source, target);
+        assertVersionNotReferenceable(target.getRoot(), path);
+    }
+
+    @Test
+    public void copyVersionTargetFrozenNodeReferenceable() throws Exception {
+        String path = "/foo";
+        NodeStore source = createVersionFor(path, createStore(true));
+        NodeStore target = createStore(true);
+
+        copyContent(source, target, path);
+        copyVersionStorage(source, target);
+        assertVersionReferenceable(target.getRoot(), path);
+    }
+
+    @Ignore("VersionCopier does not generate referenceable target nt:frozenNode from source frozen nodes that are not referenceable")
```

Comment:
The initial idea was more like test driven design. Ideally, it would be possible to copy frozen nodes that are not referenceable to a target store that has referenceable frozen nodes and the operation would generate a jcr:uuid property for each frozen node on the target store. This is currently not the case and the test is ignored. It something like a TODO, but I don't consider it a priority.